src/async_client.rs

```diff
@@ -333,6 +343,189 @@ impl AsyncClient {
+        room_id: &RoomId,
+        user_id: &UserId,
+    ) -> Result<invite_user::Response> {
+        let request = invite_user::Request {
+            room_id: room_id.clone(),
+            recipient: InvitationRecipient::UserId {
+                user_id: user_id.clone(),
+            },
+        };
+        self.send(request).await
+    }
+
+    /// Invite the specified user by third party id to the given room.
+    ///
+    /// Returns a `invite_user::Response`, an empty response.
+    ///
+    /// # Arguments
+    ///
+    /// * room_id - A valid `RoomId`.
+    ///
+    /// * invite_id - A valid `UserId`.
+    ///
+    pub async fn invite_user_by_3pid(
+        &mut self,
+        room_id: &RoomId,
+        invite_id: &Invite3pid,
+    ) -> Result<invite_user::Response> {
+        let request = invite_user::Request {
+            room_id: room_id.clone(),
+            recipient: InvitationRecipient::ThirdPartyId(invite_id.clone()),
+        };
+        self.send(request).await
+    }
+
+    /// A builder to create a room and send the request.
+    ///
+    /// Returns a `create_room::Response`, an empty response.
+    ///
+    /// # Arguments
+    ///
+    /// * room - the easiest way to create this request is using the `RoomBuilder`.
+    ///
+    /// # Examples
+    /// ```ignore
+    /// use matrix_sdk::{AsyncClient, RoomBuilder};
+    ///
+    /// let mut bldr = RoomBuilder::default();
+    /// bldr.creation_content(false)
+    ///     .initial_state(vec![])
+    ///     .visibility(Visibility::Public)
+    ///     .name("name")
+    ///     .room_version("v1.0");
+    ///
+    /// let mut cli = AsyncClient::new(homeserver, Some(session)).unwrap();
+    /// assert!(cli.create_room(bldr).await.is_ok());
+    /// ```
+    ///
+    pub async fn create_room<R: Into<create_room::Request>>(
+        &mut self,
+        room: R,
+    ) -> Result<create_room::Response> {
+        let request = room.into();
+        self.send(request).await
+    }
+
+    /// Invite the specified user by third party id to the given room.
+    ///
+    /// Returns a `invite_user::Response`, an empty response.
+    ///
+    /// # Arguments
+    ///
+    /// * request - The easiest way to create a `Request` is using the `GetMessageBuilder`
+    pub async fn room_messages<R: Into<get_message_events::Request>>(
+        &mut self,
+        request: R,
+    ) -> Result<get_message_events::IncomingResponse> {
+        let req = request.into();
+        let room_id = req.room_id.clone();
+        let mut res = self.send(req).await?;
+        let mut client = self.base_client.write().await;
+        // TODO should we support this event? to keep emitting these msg events this is needed
+        for mut event in &mut res.chunk {
+            client
+                .receive_joined_timeline_event(&room_id, &mut event)
+                .await;
+
+            if let EventResult::Ok(e) = event {
+                client.emit_timeline_event(&room_id, e).await;
```

Comment:
I think we should only emit callbacks for live events coming in with a sync. 

Do you need this for your client? I found it a bit non-intuitive that the js-sdk handles those event sources in the same way.

If we want this we'll need to have a mechanism to distinguish live events from historic ones in the callbacks.

Reply:
Hmm I don't think I need it, I think I could handle it in a different way, as you pointed out in the first review it does create some real problems doing it this way. Just return the `Response` and let the client deal with?

This may somewhat relate to the `StateStore`, will that emit events to `EventEmitter` as it could play them in the correct order. Can StateStore just store `ruma` types?

Reply:
I removed all the passing events to the `Async/BaseClient` and just return the `Response`

Reply:
> Hmm I don't think I need it, I think I could handle it in a different way, as you pointed out in the first review it does create some real problems doing it this way. Just return the `Response` and let the client deal with?

Yeah, that's how I've done it in the Python weechat-matrix as well. Seemed to work fine.
> 
> This may somewhat relate to the `StateStore`, will that emit events to `EventEmitter` as it could play them in the correct order. Can StateStore just store `ruma` types?

The most prominent state that we want to set up after a client loads is the member list. Now if we emit events when they get loaded from the state store clients won't have to implement additional logic to load the state. This sounds quite tempting at first.

The problem is clients will want to lazy load the members e.g. populate only the UI for the room they are displaying. So they would need to either 

* Only listen for events for that specific room, and implement logic to load them after the events are emitted.

* Ignore the events and implement logic to load them manually.

Seems like we can skip emitting events if they are loaded from the disk. The logic to populate UI events from a loaded client instance should be easy to implement anyways and most likely needed even if events are emitted.

Sounds reasonable or am I missing something? Would not emitting events in that case make your life unnecessarily hard?

Reply:
> Would not emitting events in that case make your life unnecessarily hard?

Yeah I wasn't thinking about it like that but that makes total sense. To make sure I'm understanding correctly: The idea is the users client like the `Async/BaseClient` will separate emitted events from "requested"/StateStore events, having 2 different mechanisms for dealing with each. I do agree that having the state store emit events would probably end up being much more work for the end client.

This makes me think the emitted events are pushed and the state stored is pulled. Setting up some way to query the state store either through the trait or another struct could be interesting? You ask by event type and one of the id's 
```rust
pub enum IdFilter {
    Room(RoomId),
    User(UserId),
    ..
}
pub fn get_event(
    &self,
    ev_type: ruma_events::EventType,
    id_filter: IdFilter
) -> Vec<ruma_events::Event> {...
```

Reply:
I think you misunderstood a bit what I meant. The state store will restore our client state in the `BaseClient` and client authors can just get the latest state that they want to present from the client object. No need to ask the state store for it.

But of course for custom setups might want to pull stuff directly from the state store so the pull setup should be possible as well.

As for the fetching specific events from the store by id, not sure if we need this or if exposed `load()`/`store()` methods are enough. Lets stick to a simple load/store and see if the need arises.